File: python/dsl/v1_0/constraint.py
class Constraint:
    """
    NDH Constraint semantic wrapper.
    Encodes altitude bounds and drift-neutrality.
    """

    def __init__(
        self,
        altitude_min: int = 4,
        altitude_max: int = 7,
        drift_neutral: bool = True,
        reversible: bool = True,
    ):
        self.altitude_min = altitude_min
        self.altitude_max = altitude_max
        self.drift_neutral = drift_neutral
        self.reversible = reversible

    def validate_altitude(self, altitude: int) -> bool:
        return self.altitude_min <= altitude <= self.altitude_max

File: rust/v1.0/Rust-StructuralRuntime-v1_0/src/constraint.rs
#[derive(Debug, Clone)]
pub struct ConstraintEnvelope {
    pub altitude_min: u8,
    pub altitude_max: u8,
    pub drift_neutral: bool,
    pub reversible: bool,
}

impl ConstraintEnvelope {
    pub fn new() -> Self {
        Self {
            altitude_min: 4,
            altitude_max: 7,
            drift_neutral: true,
            reversible: true,
        }
    }

    pub fn validate_altitude(&self, altitude: u8) -> bool {
        altitude >= self.altitude_min && altitude <= self.altitude_max
    }
}
